src/Compilers/Core/RebuildTest/RoundTripUtil.cs

```diff
@@ -86,14 +91,118 @@ public static void VerifyRoundTrip<TCompilation>(TCompilation original)
+            AssertImagesEqual(originalEmit, rebuildEmit);
+
+            unsafe EmitInfo emitOriginal()
+            {
+                switch (emitOptions.DebugInformationFormat)
+                {
+                    case DebugInformationFormat.Embedded:
+                        {
+                            var originalBytes = original.EmitToArray(emitOptions);
+                            var originalReader = new PEReader(originalBytes);
+                            var originalPdbReader = originalReader.GetEmbeddedPdbMetadataReader();
+                            Assert.NotNull(originalPdbReader);
+
+                            var pdbSpan = new ReadOnlySpan<byte>(originalPdbReader!.MetadataPointer, originalPdbReader.MetadataLength);
+                            return new EmitInfo(originalBytes, originalReader, pdbSpan.ToArray().ToImmutableArray(), originalPdbReader);
+                        }
+                    case DebugInformationFormat.PortablePdb:
+                        {
+                            using var pdbStream = new MemoryStream();
+                            var originalBytes = original.EmitToArray(emitOptions, pdbStream: pdbStream);
+                            var originalReader = new PEReader(originalBytes);
+
+                            var pdbBytes = pdbStream.ToImmutable();
+                            var originalPdbReader = MetadataReaderProvider.FromPortablePdbImage(pdbBytes).GetMetadataReader();
+                            return new EmitInfo(originalBytes, originalReader, pdbBytes, originalPdbReader);
+                        }
+                    default:
+                        throw new ArgumentException("Unsupported DebugInformationFormat: " + emitOptions.DebugInformationFormat);
+                }
+            }
+        }
+
+        private static void AssertImagesEqual(
+            EmitInfo originalEmit,
+            EmitInfo rebuildEmit)
+        {
+            if (!originalEmit.PEBytes.SequenceEqual(rebuildEmit.PEBytes))
+            {
+                var originalMdv = getMdv(originalEmit.PEReader.GetMetadataReader());
+                var rebuildMdv = getMdv(rebuildEmit.PEReader.GetMetadataReader());
+
+                // At this point the bytes were not equal, so the MDV output should also not be equal.
+                Assert.NotEqual(originalMdv, rebuildMdv);
+
+                // https://github.com/dotnet/roslyn/issues/52327
+                // this is not all that useful without manual copy/pasting. Can we diff this during the test and show the differences?
+                Assert.True(false, $@"
+Expected:
+{originalMdv}
+
+Actual:
+{rebuildMdv}
+");
+            }
+
+            if (!originalEmit.PdbBytes.SequenceEqual(rebuildEmit.PdbBytes))
+            {
+                var originalMdv = getMdv(originalEmit.PdbReader);
+                var rebuildMdv = getMdv(rebuildEmit.PdbReader);
+
+                Assert.True(false, $@"
```

Comment:
I think we should be passing this to the xUnit output infrastructure. That is usually presented with full information where as `Assert` failures will often clip information and make it hard to read.

Reply:
I will punt this to #52327 if that's ok. I agree that the whole output should probably go through the test output helper. But it might make sense to put the diff in the assert failure message.

Reply:
That's fine